libbpf-cargo/src/btf/btf.rs

```diff
@@ -0,0 +1,739 @@
+use std::cmp::{max, min};
+use std::collections::BTreeSet;
+use std::convert::TryFrom;
+use std::ffi::{c_void, CStr, CString};
+use std::fmt::Write;
+use std::mem::size_of;
+use std::slice;
+
+use anyhow::{bail, ensure, Result};
+use scroll::Pread;
+
+use crate::btf::c_types::*;
+use crate::btf::*;
+
+pub struct Btf<'a> {
+    types: Vec<BtfType<'a>>,
+    ptr_size: u32,
+    string_table: &'a [u8],
+    bpf_obj: *mut libbpf_sys::bpf_object,
+}
+
+impl<'a> Btf<'a> {
+    pub fn new(name: &str, object_file: &[u8]) -> Result<Option<Self>> {
+        let cname = CString::new(name)?;
+        let mut obj_opts = libbpf_sys::bpf_object_open_opts::default();
+        obj_opts.sz = std::mem::size_of::<libbpf_sys::bpf_object_open_opts>() as libbpf_sys::size_t;
+        obj_opts.object_name = cname.as_ptr();
+        let bpf_obj = unsafe {
+            libbpf_sys::bpf_object__open_mem(
+                object_file.as_ptr() as *const c_void,
+                object_file.len() as u64,
+                &obj_opts,
+            )
+        };
+        if bpf_obj.is_null() {
+            bail!("Failed to bpf_object__open_mem()");
+        }
+
+        let bpf_obj_btf = unsafe { libbpf_sys::bpf_object__btf(bpf_obj) };
+        if bpf_obj_btf.is_null() {
```

Comment:
similarly as for bpf_object, NULL is never returned on error (I want to change that in 1.0, but you need to use libbpf_get_error() for now)

Reply:
Does bpf_object__btf() return null on errors? https://github.com/libbpf/libbpf/blob/e05f9be4f4bed72cd2c4942bab972da8d693df08/src/libbpf.c#L8172-L8175

If we used libbpf_get_error() on a null, wouldn't it return no error?

Reply:
oops, I misread bpf_object__btf as `btf__parse`/`btf__new` constructors. Ignore this one, it will return NULL, if there is no BTF.